internal/worker/worker.go

```diff
@@ -171,177 +147,123 @@ func (worker *Worker) Run(ctx context.Context) error {
 	subCtx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	connCancel, err := worker.initializeConnection(subCtx)
-	if err != nil {
-		worker.logger.Errorf("failed to dial %s: %v", worker.rpcEndpoint, err)
-		return err
-	}
-	defer connCancel()
-
 	for {
-		if worker.sessionToken == "" {
-			if err := worker.register(subCtx); err != nil {
-				worker.logger.Errorf("failed to register worker: %v", err)
-			}
-		} else {
-			err := worker.poll(subCtx)
+		for _, upstream := range worker.upstreams {
+			if err := worker.pollSingleUpstream(subCtx, upstream); err != nil {
+				if errors.Is(err, ErrShutdown) {
+					return nil
+				}
 
-			if errors.Is(err, ErrShutdown) {
-				return nil
-			}
-
-			if err != nil {
-				worker.logger.Errorf("failed to poll: %v", err)
+				return err
 			}
 		}
 
 		select {
 		case <-subCtx.Done():
 			return nil
-		case <-time.After(time.Duration(worker.pollIntervalSeconds) * time.Second):
+		case <-time.After(time.Duration(worker.pollIntervalSeconds()) * time.Second):
 			// continue the loop
 		}
 	}
 }
 
-func (worker *Worker) initializeConnection(subCtx context.Context) (func(), error) {
-	var rpcSecurity grpc.DialOption
-
-	if worker.rpcInsecure {
-		rpcSecurity = grpc.WithTransportCredentials(insecure.NewCredentials())
-	} else {
-		certPool, _ := gocertifi.CACerts()
-		tlsCredentials := credentials.NewTLS(&tls.Config{
-			MinVersion: tls.VersionTLS13,
-			RootCAs:    certPool,
-		})
-		rpcSecurity = grpc.WithTransportCredentials(tlsCredentials)
-	}
-
-	// https://github.com/grpc/grpc-go/blob/master/Documentation/concurrency.md
-	conn, err := grpc.DialContext(subCtx, worker.rpcTarget, rpcSecurity)
-	if err == nil {
-		worker.rpcClient = api.NewCirrusWorkersServiceClient(conn)
-	}
-	return func() {
-		_ = conn.Close()
-	}, err
-}
+func (worker *Worker) pollSingleUpstream(ctx context.Context, upstream *upstreampkg.Upstream) error {
+	if err := upstream.Register(ctx, worker.info(upstream.WorkerName())); err != nil {
+		worker.logger.Errorf("failed to register worker with the upstream %s: %v",
+			upstream.Name(), err)
 
-func (worker *Worker) register(ctx context.Context) error {
-	response, err := worker.rpcClient.Register(ctx, &api.RegisterRequest{
-		WorkerInfo:        worker.info(),
-		RegistrationToken: worker.registrationToken,
-	})
-	if err != nil {
-		return err
+		return nil
 	}
 
-	worker.sessionToken = response.SessionToken
-
-	worker.logger.Infof("worker successfully registered")
-
-	return nil
-}
-
-func (worker *Worker) poll(ctx context.Context) error {
 	// De-register completed tasks
 	worker.registerTaskCompletions()
 
-	worker.logger.Debugf("polling %s", worker.rpcEndpoint)
-
 	request := &api.PollRequest{
-		WorkerInfo:   worker.info(),
-		RunningTasks: worker.runningTasks(),
+		WorkerInfo:     worker.info(upstream.WorkerName()),
+		RunningTasks:   worker.runningTasks(),
+		ResourcesInUse: worker.resourcesInUse(),
```

Comment:
I think the issue here is that tasks from all upstreams will be reported to every single upstream. I think initially we talked about making it a bit smarter and reporting only tasks polled from that particular upstream but resources in use from all the tasks of all upstreams.

I wonder if that make sense to try extend the newly created upstream logic or try to implement as we brainstormed before via leaving `Worker` as it was and introducing a new abstraction like `WorkerGroup` where the [original for-loop polling logic](https://github.com/cirruslabs/cirrus-cli/blob/f4ef224a7d730408cf8acb3170da230c39aac2ac/internal/worker/worker.go#L181-L204) will live. The logic in the worker group can look like this:

```golang
for {
  for _, worker := range group.workers {
    if worker.sessionToken == "" {
      if err := worker.register(subCtx); err != nil {
        worker.logger.Errorf("failed to register worker: %v", err)
      }
    }

    err := worker.poll(subCtx, group.resourcesInUse())
    if errors.Is(err, ErrShutdown) {
      return nil
    }
  }
   // the rest of waiting before next poll
}
```

This way the existing logic will be almost untouched. 🤔

Reply:
> I think the issue here is that tasks from all upstreams will be reported to every single upstream. I think initially we talked about making it a bit smarter and reporting only tasks polled from that particular upstream but resources in use from all the tasks of all upstreams.

Good catch, fixed this in https://github.com/cirruslabs/cirrus-cli/pull/565/commits/0ebeb27313b1739911cd3bb5442c5542a301c19b.

> I wonder if that make sense to try extend the newly created upstream logic or try to implement as we brainstormed before via leaving `Worker` as it was and introducing a new abstraction like `WorkerGroup` where the [original for-loop polling logic](https://github.com/cirruslabs/cirrus-cli/blob/f4ef224a7d730408cf8acb3170da230c39aac2ac/internal/worker/worker.go#L181-L204) will live. The logic in the worker group can look like this:

The current logic looks almost like the example you've showed above, with the exception of the registration step hidden in `worker.pollSingleUpstream()` (otherwise the `WorkerGroup` is essentially the same thing as `Upstream`, just differently named):

```go
for {
  for _, upstream := range worker.upstreams {
    if err := worker.pollSingleUpstream(subCtx, upstream); err != nil {
      if errors.Is(err, ErrShutdown) {
        return nil
      }

      worker.logger.Errorf("failed to poll upstream %s: %v", upstream.Name(), err)
    }
  }

  // the rest of waiting before next poll
}
```

I can move the `Register()` back there, but it further complicates the already nested loop (even the linter complains).

Also, are you sure we need to `group.resourcesInUse()`? Because this will hide the overall resources from other upstreams and they'll mistakenly schedule tasks to the worker that won't be scheduled if we show the full view of the worker.